### ISSUE TITLE
singleRequest to request HTTPS page does not work anymore (2.0-M2 -> 2.0) 

### ISSUE THREAD

**Issue by pme123** [akka/akka#19287]
In 2.0-M2 I used the following snippet which worked perfectly:
    `Source(Http().singleRequest(request))`

In 2.0.1 I changed it to:
    `Source.fromFuture(Http().singleRequest(request))`

I get now the following exception:
```
Caused by: javax.net.ssl.SSLException: Received fatal alert: handshake_failure
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1634)
	at sun.security.ssl.SSLEngineImpl.recvAlert(SSLEngineImpl.java:1800)
	at sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:1083)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:907)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at akka.stream.impl.io.SslTlsCipherActor.akka$stream$impl$io$SslTlsCipherActor$$doUnwrap(SslTlsCipherActor.scala:381)
	at akka.stream.impl.io.SslTlsCipherActor.akka$stream$impl$io$SslTlsCipherActor$$doInbound(SslTlsCipherActor.scala:304)
	at akka.stream.impl.io.SslTlsCipherActor$$anonfun$1.apply$mcV$sp(SslTlsCipherActor.scala:240)
	at akka.stream.impl.Pump$class.pump(Transfer.scala:199)
	at akka.stream.impl.io.SslTlsCipherActor.pump(SslTlsCipherActor.scala:45)
	at akka.stream.impl.BatchingInputBuffer.enqueueInputElement(ActorProcessor.scala:90)
	at akka.stream.impl.BatchingInputBuffer$$anonfun$upstreamRunning$1.applyOrElse(ActorProcessor.scala:141)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:16)
	at akka.stream.impl.FanIn$InputBunch$$anonfun$subreceive$1.applyOrElse(FanIn.scala:234)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:16)
	at akka.stream.impl.SubReceive.apply(Transfer.scala:12)
	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:123)
	at akka.stream.impl.SubReceive.applyOrElse(Transfer.scala:12)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:467)
	at akka.stream.impl.io.SslTlsCipherActor.aroundReceive(SslTlsCipherActor.scala:45)
```
The request retrieves a JSON over an HTTPS uri.
I found nothing in the documentation that this functionality has changed.

**Comment by ktoso** [comment]
To which URL is the problem occurring? Is it a public site? Are you _sure_ its TLS cerficicates are valid?

In 2.0 we introduced secure settings for our default HTTPS connections,
basing the code on Play's WS code which was very stable and optimised for security of outgoing HTTPS connections. PR is here: https://github.com/akka/akka/pull/19219 In other words - perhaps previously it worked, but was unsafe?

Please share more details so we can look into it.

**Comment by pme123** [comment]
Here is the URL: https://denner-api.detailnet.ch/screen/screen-publications
It can be opened in a browser without problems.

**Comment by johanandren** [comment]
I looked a bit at this, cert and chain looks ok, root cert is in some trust store at least on my computer. Works using `URL.openConnection`, but not using `Http().singleRequest`

Not sure if this is relevant but with a breakpoint in SslTlsCipherActor.fail i can see the `handshake_failure` from above interleaved with this:
```
Inbound closed before receiving peer's close_notify: possible truncation attack?
```
I guess that might be because I'm breakpointing though.

**Comment by ktoso** [comment]
did you run `java` with `-Djavax.net.debug=all`? It should include info why it rejects.

**Comment by johanandren** [comment]
Now I did:
[output.txt](https://github.com/akka/akka/files/81418/output.txt)

**Comment by ktoso** [comment]
Thanks! I'll leave that for tomorrow, now must do slides :~

The important bit is (perhaps it simply does not support any of the encryption schemes we allow by default – see ssl-configu defaults to see which ones we support – as the number of ignoring unsupported cipher suite is quite large hm):


```
Ignoring unsupported cipher suite: TLS_DHE_DSS_WITH_AES_256_GCM_SHA384
...
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
...
*** ClientHello, TLSv1
Session ID:  {}
Cipher Suites: [TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, ...TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, 
TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA, TLS_ECDH_RSA_WITH_AES_256_CBC_SHA, TLS_DHE_RSA_WITH_AES_256_CBC_SHA, TLS_DHE_DSS_WITH_AES_256_CBC_SHA, 
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA, 
TLS_ECDH_RSA_WITH_AES_128_CBC_SHA, TLS_DHE_RSA_WITH_AES_128_CBC_SHA, TLS_DHE_DSS_WITH_AES_128_CBC_SHA, TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA, 
TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA, SSL_RSA_WITH_3DES_EDE_CBC_SHA, TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA, TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA, 
SSL_DHE_RSA_WITH_3DES_EDE_CBC_SHA, SSL_DHE_DSS_WITH_3DES_EDE_CBC_SHA, TLS_EMPTY_RENEGOTIATION_INFO_SCSV]
Compression Methods:  { 0 }
Extension elliptic_curves, curve names: {secp256r1, sect163k1, sect163r2, secp192r1, secp224r1, sect233k1, sect233r1, sect283k1, sect283r1, secp384r1, sect409k1, sect409r1, secp521r1, sect571k1, sect571r1, secp160k1, secp160r1, secp160r2, sect163r1, secp192k1, sect193r1, sect193r2, secp224k1, sect239k1, secp256k1}
Extension ec_point_formats, formats: [uncompressed]
***...
default-akka.actor.default-dispatcher-3, RECV TLSv1.2 ALERT:  fatal, handshake_failure
default-akka.actor.default-dispatcher-3, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Received fatal alert: handshake_failure
default-akka.actor.default-dispatcher-3, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Received fatal alert: handshake_failure
default-akka.actor.default-dispatcher-3, called closeInbound()
default-akka.actor.default-dispatcher-3, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
[DEBUG] [01/07/2016 17:49:02.544] [default-akka.actor.default-dispatcher-3] [akka://default/user/StreamSupervisor-0/flow-1-2-SlotProcessorInternalConnectionFlow] closing output
[DEBUG] [01/07/2016 17:49:02.547] [default-akka.actor.default-dispatcher-4] [akka://default/user/SlotProcessor-0] Slot 0 disconnected after javax.net.ssl.SSLException: Received fatal alert: handshake_failure
```

We should check if the supported protocols overlap with what ssl-config allows - could you check that? Let me know if more hints needed.

FYI @wsargent

**Comment by jrudolph** [comment]
Btw, this still works on one of my Javas:

```
java version "1.7.0_91"
OpenJDK Runtime Environment (IcedTea 2.6.3) (7u91-2.6.3-0ubuntu0.14.04.1)
OpenJDK 64-Bit Server VM (build 24.91-b01, mixed mode)
```

but fails on Java 8.

**Comment by jrudolph** [comment]
It seems that Java 8 is not sending the SNI extension. Maybe it's this: http://stackoverflow.com/questions/30817934/extended-server-name-sni-extension-not-sent-with-jdk1-8-0-but-send-with-jdk1-7

**Comment by jrudolph** [comment]
The problem seems to be that Java 8 doesn't enable the SNI extension automatically any more. Instead, code like shown in the JSSE documentation is needed:

https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/JSSERefGuide.html#SNIExtension

**Comment by ktoso** [comment]
Thanks a lot for finding that! Sounds like an addition to be made in ssl-config.

**Comment by wsargent** [comment]
Also see https://bugs.openjdk.java.net/browse/JDK-8144566 and it looks like some extra code is needed :-(

**Comment by jrudolph** [comment]
This JDK issue and the issue linked above in the SO question seem a bit bogus. I previously thought they were related as well but the scenario is actually different because they don't seem to rely on `SSLEngine` directly.

I forgot about this issue but I think I hacked the solution given in the JSSE documentation directly into the SslTlsStage and it seemed to work.

This fix also needs to be backported to stream-http-2.0.

**Comment by pme123** [comment]
Will this fix be in the next stream-http-2.0 release and if when will that be?

**Comment by samueltardieu** [comment]
@jrudolph I've been hit by the very same SNI client problem on 2.4.2-RC2. Is there a chance the fix will go in 2.4.2?

**Comment by nktpro** [comment]
I also encountered this exact issue today with `Http().singleRequest(...)` to AWS API Gateway endpoints. Is there a reasonable workaround meanwhile?

**Comment by pme123** [comment]
If you need it urgent: I switched to commons apache-http.

**Comment by jrudolph** [comment]
I think this should be fixed very soon. It is in so far a regression that running with Java 6 used to work but now doesn't because Akka 2.4 requires Java 8.

**Comment by ktoso** [comment]
It's quite on the top of my list, will work on it soon.

**Comment by ktoso** [comment]
Solved, will be shipped in 2.4.3.

**Comment by ktoso** [comment]
This likely solved https://github.com/akka/akka/issues/16796
